core/arch/x86/mangle.c

```diff
@@ -2391,6 +2394,31 @@ mangle_interrupt(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
 }
 
 /***************************************************************************
+ * Single step exceptions generation
+ */
+void
+mangle_single_step(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr)
+{
+    /*
+     * Simply inserts a nop so as that next instruction where a single step
+     * exception might occur is not in a different basic block.
+     * FIXME i#2144 : to be absolutely transparent, we should translate the
+     * exception address as if we did not insert this nop.
+     */
+    instr_t* next = instr_get_next_app(instr);
+    /* Assumes popf cannot end a basic block. */
+    ASSERT(next);
+    /*
+     * The single step exception is only a problem on a control transfer
+     * because the ExceptionAddress should be the jump target.
+     */
+    if (instr_is_cti(next)) {
+        /* Next app instruction gives a correct translation to this nop. */
```

Comment:
mangle_shared.c:865:
```
            app_pc xl8 = get_app_instr_xl8(instr);
            instrlist_set_translation_target(ilist, xl8);
```

It doesn't matter *where* the nop is inserted: its xl8 will be set to ilist->translation_target, which is the xl8 of the instr being mangled, which is the popf.